### DISCUSSION TITLE
Connection pooling for multithread apps

### DISCUSSION THREAD

**Discussion by hmmftg** [sijms/go-ora#72] [Q&A] [answered]
Hi, I'm writing a multithread(go routine) application, and it is newly started

It has a producer consumer approach to handle some tcp-ip requests, and requests are very much(near 1500 requests per second), they should call some plsql functions per request. and this was done by C and we are deciding to migrate to GO

Our server is AIX and the only go oracle driver which supports cross-compilation for our environment is your great package

We used PRO*C pre-compiler provided by oracle in our C codes, it supports multithreads and we implement a connection pool to use

My question is: can we use your package and don't worry about concurrencies?

**Comment by sijms** [comment]
yes of course 
but you need to make one connection per thread (go routine) and avoid sharing one connection with multiple thread unless you use anti-race condition measure like mutex

the only difference between this method and connection pooling is that connection pooling preserve a number of open connection (predefined) during program execution then when you call open you take control of one connection until you release it by close and so on thus the time of open and close connection is positioned at the beginning and at the end of the program, 
while the above method you open connection with the thread start and close it with thread termination (open connection make slight delay when thread start) which will also occur with the pooling during pool grow

**Comment by sijms** [comment, marked as answer]
it also seem that the package database/sql internally support connection pooling thus you can use the connection object safely between threads in multi-thread application

see this code from official Golang package site [here](https://golang.org/pkg/database/sql/driver/)
```
type Connector interface {
    // Connect returns a connection to the database.
    // Connect may return a cached connection (one previously
    // closed), but doing so is unnecessary; the sql package
    // maintains a pool of idle connections for efficient re-use.
    //
    // The provided context.Context is for dialing purposes only
    // (see net.DialContext) and should not be stored or used for
    // other purposes. A default timeout should still be used
    // when dialing as a connection pool may call Connect
    // asynchronously to any query.
    //
    // The returned connection is only used by one goroutine at a
    // time.
    Connect(context.Context) (Conn, error)

    // Driver returns the underlying Driver of the Connector,
    // mainly to maintain compatibility with the Driver method
    // on sql.DB.
    Driver() Driver
}
```

**Reply by hmmftg** [reply]
Thank you for your reply
